boot/freeldr/freeldr/arch/uefi/uefimem.c

```diff
@@ -0,0 +1,309 @@
+            return LoaderReserve;
+        case EfiLoaderCode:
+            return LoaderLoadedProgram;
+        case EfiLoaderData:
+            return LoaderLoadedProgram;
+        case EfiBootServicesCode:
+            return LoaderFirmwareTemporary;
+        case EfiBootServicesData:
+            return LoaderFirmwareTemporary;
+        case EfiRuntimeServicesCode:
+            return LoaderFirmwarePermanent;
+        case EfiRuntimeServicesData:
+            return LoaderFirmwarePermanent;
+        case EfiConventionalMemory:
+            return LoaderFree;
+        case EfiUnusableMemory:
+            return LoaderBad;
+        case EfiACPIReclaimMemory:
+            return LoaderFirmwareTemporary;
+        case EfiACPIMemoryNVS:
+            return LoaderReserve;
+        case EfiMemoryMappedIO:
+            return LoaderReserve;
+        case EfiMemoryMappedIOPortSpace:
+            return LoaderReserve;
+        default:
+            break;
+    }
+    return LoaderReserve;
+}
+
+PFREELDR_MEMORY_DESCRIPTOR
+UefiMemGetMemoryMap(ULONG *MemoryMapSize)
+{
+    EFI_LOADED_IMAGE_PROTOCOL *LoadedImage;
+    UINT32 DescriptorVersion;
+    SIZE_T FreeldrMemMapSize;
+    UINTN DescriptorSize;
+    EFI_STATUS Status;
+    UINTN MapSize;
+    UINTN MapKey;
+    UINT32 Index;
+
+    EFI_GUID EfiLoadedImageProtocol = EFI_LOADED_IMAGE_PROTOCOL_GUID;
+    PFREELDR_MEMORY_DESCRIPTOR FreeldrMem = NULL;
+    EFI_MEMORY_DESCRIPTOR* MapEntry = NULL;
+    UINT32 EntryCount = 0;
+    FreeldrDescCount = 0;
+
+    Status = GlobalSystemTable->BootServices->HandleProtocol(GlobalImageHandle,
+                                                             &EfiLoadedImageProtocol,
+                                                             (VOID**)&LoadedImage);
+    if (Status != EFI_SUCCESS)
+    {
+        TRACE("Failed to find LoadedImageHandlewith status: %d", Status);
+        UiMessageBoxCritical("Unable to initialize memory manager.");
+    }
+    OsLoaderBase = LoadedImage->ImageBase;
+    OsLoaderSize = LoadedImage->ImageSize;
+    PublicBootHandle = LoadedImage->DeviceHandle;
+    EfiMemoryMap = NULL;
+
+
+    TRACE("UefiMemGetMemoryMap: Gather memory map\n");
+    PUEFI_LoadMemoryMap(&MapKey,
+                        &MapSize,
+                        &DescriptorSize,
+                        &DescriptorVersion);
+
+    TRACE("Value of MapKey: %d\n", MapKey);
+    TRACE("Value of MapSize: %d\n", MapSize);
+    TRACE("Value of DescriptorSize: %d\n", DescriptorSize);
+    TRACE("Value of DescriptorVersion: %d\n", DescriptorVersion);
+
+    EntryCount = (MapSize / DescriptorSize);
+
+    FreeldrMemMapSize = (sizeof(FREELDR_MEMORY_DESCRIPTOR) * EntryCount);
+	Status = GlobalSystemTable->BootServices->AllocatePool(EfiLoaderData, FreeldrMemMapSize, (void**)&FreeldrMem);
+    if (Status != EFI_SUCCESS)
+    {
+        TRACE("Failed to allocate pool with status %d", Status);
+        UiMessageBoxCritical("Unable to initialize memory manager.");
```

Comment:
What about returning failure? How does the caller of UefiMemGetMemoryMap() handles that case?
```suggestion
        UiMessageBoxCritical("Unable to initialize memory manager.");
        return NULL;
```